### PR TITLE
Handle alternative phrasings

### PR DESCRIPTION
**Pull request by mruwnik**
https://github.com/StampyAI/stampy-nlp/issues/13

### CODE REVIEW

File: src/stampy_nlp/utilities/coda_utils.py [hunk 1]
```diff
@@ -21,21 +21,30 @@
...
-
+def fetch_all_rows():
     headers: object = {'Authorization': f'Bearer {get_coda_token()}'}
     params: object = {
         'useColumnNames': True,
         'limit': 1000
     }
-    json_items = requests.get(TABLE_URL, headers=headers, params=params).json()['items']
+    response = requests.get(TABLE_URL, headers=headers, params=params).json()
+    while response:
```
Comment (mruwnik) [inline src/stampy_nlp/utilities/coda_utils.py:31]: coda will not return more than 500 items - to get more the results have to be paginated :/

**Review by ccstan99 — COMMENTED** [review]
One potential hiccup doing it this way, we may be treating the alternate phrase on "equal footing" with the actual / original question, i.e. when the results are returned, it may not be clear what was the original title of the question and which are alternatives. I don't know if this'll be an issue. Do we need/want an additional field to keep track?

Also, just doing a quick check on a several alternate phrasing, the python sentence transformer already pulls them as most similar. Maybe the tensorflow is less accurate but still in the top 5 on several.

Finally, I'll want to test this out before approving, but I probably won't have time to do it properly until the weekend.

File: src/stampy_nlp/utilities/coda_utils.py (hunk 1 above)
Comment (mruwnik) [reply src/stampy_nlp/utilities/coda_utils.py:31]: I was getting back 500 results. Which is a suspiciously round number :D

**Review by ccstan99 — CHANGES_REQUESTED** [review]
Unfortunately, the NLP endpoint doesn't work since the alternate phrases are using the same id which must be unique in pinecone. We should probably use a hash value or something instead of using the coda id. Of course, the deduplication will need to filter by something else, like paged since the id will no longer match.

File: src/stampy_nlp/utilities/coda_utils.py [hunk 2]
```diff
@@ -45,13 +54,16 @@ def get_df_data():
...
-                'title': item['name'],
-                'status': status,
-                'pageid': pageid,
-                'url': url,
-            })
+            for i, name in enumerate(names):
+                data_list.append({
+                    # ids must be unique, and the first name is the original name, which means that
+                    # any subsequent names are duplicates
+                    'id': f'{item["id"]}{i or ""}',
```
Comment (mruwnik) [inline src/stampy_nlp/utilities/coda_utils.py:61]: a bit of magic here to make sure that the ids are unique. This has the nice effect of leaving the prexisting values in the index as are, and adding new ids for alternative phrasings

**Review by ccstan99 — APPROVED** [review]
Great! Your "magic" is a better way to preserve the existing id. I was going to suggest it after reading your earlier discord message but of course, you beat me to it. 

Since pageid is stored in pinecone metadata, the deduplication needs to check `result['metadata']['pageid']` not since `result['pageid']` isn't a valid field. I made a commit with that change.

Both pinecone and tensorflow's json now include the embeddings for the alternate phrasings in their deployed versions. However, I think something might still be off with the pinecone embeddings. There were originally 439 questions + 28 alternate phrasings so we should have 467 but there's 745 now. It doesn't seem to cause any obvious problems but needs to be further investigated. Unfortunately, I'm swamped preparing for a couple of presentations for Wednesday and Thursday.

In the meantime, feel free to merge this and the stampy-ui deduplication PR whenever you're comfortable.

**Comment by mruwnik** [comment]
That actually makes sense, since there are some 720 answers returned when pagination  is used

**Pull request merged**